grakn-engine/src/main/java/ai/grakn/engine/rpc/TxObserver.java

```diff
@@ -0,0 +1,220 @@
+
+    private final StreamObserver<TxResponse> responseObserver;
+    private final EngineGraknTxFactory txFactory;
+    private final AtomicBoolean terminated = new AtomicBoolean(false);
+
+    private @Nullable GraknTx tx = null;
+    private @Nullable Iterator<QueryResult> queryResults = null;
+
+    private static final TxResponse DONE = TxResponse.newBuilder().setDone(Done.getDefaultInstance()).build();
+
+    private TxObserver(EngineGraknTxFactory txFactory, StreamObserver<TxResponse> responseObserver) {
+        this.responseObserver = responseObserver;
+        this.txFactory = txFactory;
+    }
+
+    public static TxObserver create(EngineGraknTxFactory txFactory, StreamObserver<TxResponse> responseObserver) {
+        return new TxObserver(txFactory, responseObserver);
+    }
+
+    @Override
+    public void onNext(TxRequest request) {
+        try {
+            switch (request.getRequestCase()) {
+                case OPEN:
+                    open(request.getOpen());
+                    break;
+                case COMMIT:
+                    commit();
+                    break;
+                case EXECQUERY:
+                    execQuery(request.getExecQuery());
+                    break;
+                case NEXT:
+                    next();
+                    break;
+                case STOP:
+                    stop();
+                    break;
+                default:
+                case REQUEST_NOT_SET:
+                    throw error(Status.INVALID_ARGUMENT);
+            }
+        } catch (GraknException e) {
+            Metadata trailers = new Metadata();
+            trailers.put(GrpcServer.MESSAGE, e.getMessage());
+            throw error(Status.UNKNOWN, trailers);
+        }
+    }
+
+    private void open(Open request) {
+        if (tx != null) {
+            throw error(Status.FAILED_PRECONDITION);
+        }
+
+        String keyspaceString = request.getKeyspace().getValue();
+        Keyspace keyspace = Keyspace.of(keyspaceString);
+        GraknTxType txType = getTxType(request.getTxType());
+        tx = txFactory.tx(keyspace, txType);
+
+        responseObserver.onNext(DONE);
+    }
+
+    private void commit() {
+        if (tx == null) {
+            throw error(Status.FAILED_PRECONDITION);
+        }
+        tx.commit();
+    }
+
+    private void execQuery(ExecQuery request) {
+        if (tx == null || queryResults != null) {
+            throw error(Status.FAILED_PRECONDITION);
+        }
+
+        String queryString = request.getQuery().getValue();
+
+        QueryBuilder graql = setInfer(tx.graql(), request.getInfer());
+
+        queryResults = graql.parse(queryString).results(GrpcConverter.get()).iterator();
+
+        sendNextResult(queryResults);
+    }
+
+    private void next() {
+        if (queryResults == null) {
+            throw error(Status.FAILED_PRECONDITION);
+        }
+
+        sendNextResult(queryResults);
```

Comment:
Why do we pass around the queryResults as a parameter? Since it's already defined as a class property